### PR TITLE
BIP-X: Tractor (+ Junctions)

### PR DESCRIPTION
**Pull request by funderbrker**
## Tractor
Tractor is a system to give a third party operator permission to perform *any* Beanstalk action on behalf of the publisher.

### CODE REVIEW

**Comment by funderbrker** [comment]
## Status
- Operational and tested. All tests pass.
- The underlying branch has changed, causing conflicts. Should not affect logic though.
- SDK being implemented here https://github.com/BeanstalkFarms/Beanstalk/pull/791

**Review by Brean0 — COMMENTED** [review]
review I

File: protocol/contracts/libraries/LibTractor.sol [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    bytes32 private constant EIP712_TYPE_HASH =
+        keccak256(
+            "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+        );
+    bytes32 public constant BLUEPRINT_TYPE_HASH =
+        keccak256(
+            "Blueprint(address publisher,bytes data,bytes operatorData,uint256 maxNonce,uint256 startTime,uint256 endTime)"
+        );
+
+    struct TractorStorage {
```
[the remark this answers is not in the document]
Comment (funderbrker) [reply protocol/contracts/libraries/LibTractor.sol:33]: let's punt on this until we decide how we want to approach larger scale restructuring and storage changes

File: protocol/contracts/libraries/LibTractor.sol [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    /// @return hash calculated Blueprint hash
+    function _getBlueprintHash(Blueprint calldata blueprint) internal view returns (bytes32) {
+        return
+            _hashTypedDataV4(
+                keccak256(
+                    abi.encode(
+                        BLUEPRINT_TYPE_HASH,
+                        blueprint.publisher,
+                        keccak256(blueprint.data),
+                        keccak256(abi.encodePacked(blueprint.operatorPasteInstrs)),
```
[the remark this answers is not in the document]
Comment (funderbrker) [reply protocol/contracts/libraries/LibTractor.sol:131]: great question
it's an inline way for adhering to the eip712 spec

the correct way to pack a list of objects in eip712 is to 
1. encode each object
2. concatenate all encoded data
3. hash the concatenation

here each object is a bytes32 so it is already 'encoded', so the encodePacked directly concatenates them all without extraneous bytes. then they are hashed.

File: protocol/contracts/C.sol [hunk 3]
```diff
@@ -84,6 +84,15 @@ library C {
     uint256 internal constant BEAN_INDEX = 0;
     uint256 internal constant ETH_INDEX = 1;
 
+    //////////////////// Tractor ////////////////////
+
+    uint80 internal constant SELECTOR_SIZE = 4;
+    uint80 internal constant SLOT_SIZE = 32;
+    uint80 internal constant ARGS_START_INDEX = SELECTOR_SIZE + SLOT_SIZE;
+    uint80 internal constant ADDR_SLOT_OFFSET = 12;
+    uint80 internal constant PUBLISHER_COPY_INDEX = type(uint80).max;
```
[the remark this answers is not in the document]
Comment (funderbrker) [reply protocol/contracts/C.sol:92]: fixed
https://github.com/BeanstalkFarms/Beanstalk/pull/839/files#diff-27881ed6f4fd2273b16a73711106873d04e4eb6b8461c5356d7273a5619d4844R100

File: protocol/contracts/libraries/LibFarm.sol [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+// AdvancedFarmCall is a Farm call that can use a Clipboard.
+// See LibFunction.useClipboard for details
+struct AdvancedFarmCall {
+    bytes callData;
+    bytes clipboard;
+}
+
+library LibFarm {
+    function _advancedFarm(
+        AdvancedFarmCall calldata data,
```
[the remark this answers is not in the document]
Comment (funderbrker) [reply protocol/contracts/libraries/LibFarm.sol:27]: removed in facet and lib
https://github.com/BeanstalkFarms/Beanstalk/pull/839/files#diff-3d41dc36c1eaa68f2306e7d5d32a34c9b88bb89d52714afde9827cae0d10876fL27

File: protocol/contracts/libraries/LibTractor.sol [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        // keccak256("diamond.storage.tractor") == 0x7efbaaac9214ca1879e26b4df38e29a72561affb741bba775ce66d5bb6a82a07
+        assembly {
+            ts.slot := 0x7efbaaac9214ca1879e26b4df38e29a72561affb741bba775ce66d5bb6a82a07
+        }
+    }
+
+    /// @notice increment the blueprint nonce by 1
+    /// @param blueprintHash blueprint hash
+    function _incrementBlueprintNonce(bytes32 blueprintHash) internal {
+        TractorStorage storage ts = _tractorStorage();
```
[the remark this answers is not in the document]
Comment (funderbrker) [reply protocol/contracts/libraries/LibTractor.sol:71]: fixed 
https://github.com/BeanstalkFarms/Beanstalk/pull/839/files#diff-a9cf44d910ac787cc810d4580304c10cd0913b0ef3dcb38ba21134772d7e5e46R77

File: protocol/contracts/libraries/LibTractor.sol [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    /// @notice set blueprint publisher address
+    /// @param publisher blueprint publisher address
+    function _setPublisher(address publisher) internal {
+        require(
+            _tractorStorage().activePublisher == address(1),
+            "LibTractor: publisher already set"
+        );
+        _tractorStorage().activePublisher = publisher;
```
[the remark this answers is not in the document]
Comment (funderbrker) [reply protocol/contracts/libraries/LibTractor.sol:89]: probably. anyway it follows the style we use for app storage. updated here
https://github.com/BeanstalkFarms/Beanstalk/pull/839/files#diff-a9cf44d910ac787cc810d4580304c10cd0913b0ef3dcb38ba21134772d7e5e46R94

**Pull request merged**